boot/freeldr/freeldr/lib/fs/iso.c

```diff
@@ -140,116 +169,148 @@ static ARC_STATUS IsoBufferDirectory(ULONG DeviceId, ULONG DirectoryStartSector,
 
-    //
-    // Read the Primary Volume Descriptor
-    //
-    Position.HighPart = 0;
-    Position.LowPart = 16 * SECTORSIZE;
-    Status = ArcSeek(DeviceId, &Position, SeekAbsolute);
-    if (Status != ESUCCESS)
-        return Status;
-    Status = ArcRead(DeviceId, Pvd, SECTORSIZE, &Count);
-    if (Status != ESUCCESS || Count < sizeof(PVD))
-        return EIO;
+    /*
+     * Extract the file name
+     * '\test.ini' --> 'test.ini'
+     * '\folder\app.exe' --> 'app.exe'
+     */
+    FileNameStr = IsoLastPathSeparator(FileName) + 1;
 
-    DirectorySector = Pvd->RootDirRecord.ExtentLocationL;
-    DirectoryLength = Pvd->RootDirRecord.DataLengthL;
+    /*
+     * Extract the directory path, including trailing slash
+     * '\test.ini' --> '\'
+     * '\folder\app.exe' --> '\folder\'
+     */
+    DirectoryPathLength = FileNameStr - FileName;
 
-    /* Skip leading path separator, if any */
-    if (*FileName == '\\' || *FileName == '/')
-        ++FileName;
-    //
-    // Figure out how many sub-directories we are nested in
-    //
-    NumberOfPathParts = FsGetNumPathParts(FileName);
+    /* See if this directory has been buffered before */
+    DoFullLookup = (DirectoryPathLength != Volume->DirectoryPathLength) ||
+                   !RtlEqualMemory(FileName, Volume->DirectoryPath, DirectoryPathLength);
+    if (!DoFullLookup)
+    {
+        PathPart = (CHAR*)FileNameStr;
+        DirectoryBuffer = Volume->DirectoryBuffer;
+        DirectoryLength = Volume->DirectoryLength;
+
+        NumberOfPathParts = 1;
+    }
+    else
+    {
+        PathPart = PathBuffer;
+        DirectorySector = Volume->PvdDirectorySector;
+        DirectoryLength = Volume->PvdDirectoryLength;
+
+        /* Skip leading path separator, if any */
+        if (*FileName == '\\' || *FileName == '/')
+            ++FileName;
+
+        /* Figure out how many sub-directories we are nested in */
+        NumberOfPathParts = FsGetNumPathParts(FileName);
+    }
 
     //
     // Loop once for each part
     //
     for (i=0; i<NumberOfPathParts; i++)
     {
-        //
-        // Get first path part
-        //
-        FsGetFirstNameFromPath(PathPart, FileName);
-
-        //
-        // Advance to the next part of the path
-        //
-        for (; (*FileName != '\\') && (*FileName != '/') && (*FileName != '\0'); FileName++)
+        if (DoFullLookup)
         {
-        }
-        FileName++;
+            /* Get first path part */
+            FsGetFirstNameFromPath(PathPart, FileName);
 
-        //
-        // Buffer the directory contents
-        //
-        Status = IsoBufferDirectory(DeviceId, DirectorySector, DirectoryLength, &DirectoryBuffer);
-        if (Status != ESUCCESS)
-            return Status;
+            /* Advance to the next part of the path */
+            while ((*FileName != ANSI_NULL) && (*FileName != '\\') && (*FileName != '/'))
+            {
+                FileName++;
+            }
+            FileName++;
```

Comment:
```suggestion
            FileName++; /* Skip the path separator */
```
I understand here that this extra FileName++ would skip past the encountered `'\\'` or `'/'`. But what if the while-loop stopped because `*FileName == ANSI_NULL` ? Wouldn't this move FileName outside of the string buffer?

Reply:
You are right. Meanwhile, this has no consequence as `FileName` won't be used anymore, as the loop will exit on `NumberOfPathParts` condition.
Nonetheless, an `if (*FileName != ANSI_NULL)` could be added, to be explicit/future-proof.

--

For a separate PR,
`FsGetNumPathParts()` similar code is affected! It looks like it should be fixed (or explicitly documented) to handle an empty string, maybe even a `NULL` input too.
https://git.reactos.org/?p=reactos.git;a=blob;f=boot/freeldr/freeldr/lib/fs/fs.c;hb=5461abeeba643d6ba1cce3f0610b023f71afc1e9#l534

Reply:
> Wouldn't this move FileName outside of the string buffer?

That time it's guaranteed to be the last iteration in the loop by `FsGetNumPathParts()` called before.